### PR TITLE
Issue #12608: Validation for next version in GitHub action

### LINKED ISSUE #12608: Validation for next version in GitHub action
I did typo at version bump
https://github.com/checkstyle/checkstyle/actions/runs/3863258249/jobs/6585335155#step:3:1

I put 1.7.0 instead 10.7.0

We need enforce validation to be allow bump only for +1 in any of digits from previous. Ideally we should use more smart validation, but let's see what can we do with less efforts

### PR DESCRIPTION
**Pull request by stoyanK7**
Resolves #12608 

---
Screenshot from 2023-02-05 00-22-47

### CODE REVIEW

**Comment by stoyanK7** [comment]
@romani Is this what we are aiming for?

**Comment by rnveach** [comment]
@stoyanK7 Your screenshot is what I would ask for.

Can you also show the output when numbers are accidentally decreased, as that is what the reported issue is about.

Example: We are on 10.7.1 . What are the outputs of 1.7.1, 10.1.1, 10.7.0, etc...

Also, please confirm if we have any validation to ensure version number is in the form `\d+.\d+.\d+` and show an example?

**Comment by romani** [comment]
@stoyanK7 , better to share CLI text output, not a image.

But yes, changes are good enough. It will cover most typos, being very detailed might be very expensive in implementation.

**Comment by stoyanK7** [comment]
@rnveach I just added an assertion for `x.x.x`. Here's the output as requested:
```sh
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 1.7.1                                                                                                                                                      12:36:56 pm
NEW_VERSION=1.7.1
CURRENT_VERSION=10.7.1
[ERROR] Major number(10) has changed to 1 but it should be incremented by max 1.
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.1.1                                                                                                                                                     12:37:02 pm
NEW_VERSION=10.1.1
CURRENT_VERSION=10.7.1
[ERROR] Minor number(7) has changed to 1 but it should be incremented by max 1.
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.0                                                                                                                                                     12:37:07 pm
NEW_VERSION=10.7.0
CURRENT_VERSION=10.7.1
[ERROR] Patch number(1) has changed to 0 but it should be incremented by max 1.
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.0.0                                                                                                                                                   12:37:13 pm
NEW_VERSION=10.7.0.0
[ERROR] New version(10.7.0.0) is not in the format of <major>.<minor>.<patch>
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.-                                                                                                                                                     12:37:24 pm
NEW_VERSION=10.7.-
[ERROR] New version(10.7.-) is not in the format of <major>.<minor>.<patch>
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.0.                                                                                                                                                    12:37:32 pm
NEW_VERSION=10.7.0.
[ERROR] New version(10.7.0.) is not in the format of <major>.<minor>.<patch>

```

**Comment by stoyanK7** [comment]
Final check with valid versions:
```sh
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.1                                                                                                                                                     12:43:51 pm
NEW_VERSION=10.7.1
CURRENT_VERSION=10.7.1
[ERROR] New Version and Current Version are the same.
______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.7.2                                                                                                                                                     12:43:55 pm
NEW_VERSION=10.7.2
CURRENT_VERSION=10.7.1
bump version in pom.xml
^C                                                                                                                                                                                            ______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 10.8.0                                                                                                                                                     12:43:59 pm
NEW_VERSION=10.8.0
CURRENT_VERSION=10.7.1
bump version in pom.xml
^C                                                                                                                                                                                            ______________________________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                                issue/12608-validate-next-ver
$ ./.ci/bump-version.sh 11.0.0                                                                                                                                                     12:44:04 pm
NEW_VERSION=11.0.0
CURRENT_VERSION=10.7.1
bump version in pom.xml

```

**Review by romani — APPROVED** [review]
Ok to merge

File: .ci/bump-version.sh [hunk 1]
```diff
@@ -1,9 +1,68 @@
...
+NEW_VERSION=$1
+echo NEW_VERSION="$NEW_VERSION"
+
+if ! [[ "$NEW_VERSION" =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+  echo "[ERROR] New version($NEW_VERSION) is not in the format of <major>.<minor>.<patch>"
+  exit 1;
+fi
+
+CURRENT_VERSION=$(xmlstarlet sel -N pom=http://maven.apache.org/POM/4.0.0 \
+                             -t -m pom:project -v pom:version pom.xml | sed "s/-SNAPSHOT//")
```
Comment (rnveach) [inline .ci/bump-version.sh:19]: Why can't we replace this with call to `getCheckstylePomVersion` from util.sh?

File: .ci/bump-version.sh [hunk 2]
```diff
@@ -1,9 +1,68 @@
...
+if [[ -z $1 ]]; then
+  echo "version is not set"
+  echo "Usage: $BASH_SOURCE <version>"
+  exit 1
+fi
+
+NEW_VERSION=$1
+echo NEW_VERSION="$NEW_VERSION"
+
+if ! [[ "$NEW_VERSION" =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
```
Comment (rnveach) [inline .ci/bump-version.sh:13]: Why `[0-9]` and not `\d`?

**Review by rnveach — CHANGES_REQUESTED** [review]
Very minor changes look for.

File: .ci/bump-version.sh (hunk 2 above)
Comment (stoyanK7) [reply .ci/bump-version.sh:13]: `\d` does not work.

This answer explains why: https://stackoverflow.com/a/17968116/9553927
Screenshot from 2023-02-05 18-20-07

File: .ci/bump-version.sh (hunk 1 above)
Comment (stoyanK7) [reply .ci/bump-version.sh:19]: No issues in doing it. It just takes 1-2 seconds longer.

I saw that we have other places where this `xmlstarlet` call can be replaced with a call to new function `getCheckstylePomVersionWithoutSnapshot`. Shall I open an issue for this?

Done. Replaced.

Comment (rnveach) [reply .ci/bump-version.sh]: Please do.

**Review by rnveach — APPROVED** [review]

**Pull request merged**